### PR TITLE
task: add ways to run a `LocalSet` from within a rt context

### LINKED ISSUE #1906: task: Add some way to drive a `LocalSet` inside of a task
Currently, the only way to run a `tokio::task::LocalSet` is to call its `block_on` method with a `&mut Runtime`, like
```rust
let mut rt = tokio::runtime::Runtime::new();
let local = tokio::task::LocalSet::new();
local.block_on(&mut rt, async {
  // whatever...
});
```

Unfortunately, this means that `LocalSet` doesn't work with the `#[tokio::main]`  and `#[tokio::test]` macros, since the `main` function is _already_ inside of a call to `block_on`.

We should consider adding a function like
```rust
pub fn run_with<F: Future>(&self, f: F) -> LocalFuture<F>
```
to `LocalSet`, so that users can write code like
```rust
#[tokio::main]
async fn main() {
    let local = tokio::task::LocalSet::new();
    local.run_with(async {
         // .. whatever ...
    }).await;
}
```
We could also add a `run()` or `run_until_idle()` that just drives the local set until all tasks spawned on it complete.

### LINKED ISSUE #1908: task::LocalSet: provide `async fn await_on(impl Future)` and `async fn await_for_idle`
naming TBD:

Providing such functions would allow using `LocalSet` directly from `#[tokio::main`].

* `await_on` would take a future to await.
* `await_for_idle` would run until the `LocalSet` is idle (no spawned futures).

### LINKED ISSUE #2057: reqwest stalls in locally-spawned task when "full" feature is not selected
## Version
```
[dependencies]
reqwest = "0.10"
tokio = { version = "0.2", features = ["net", "rt-util"] }
```

## Platform
```
Darwin niobium-2.local 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
```

(I've also seen the below-described behavior on Linux.)

## Description

```rust
use tokio::sync::oneshot::channel;

fn main() {
    let mut rt = tokio::runtime::Runtime::new().unwrap();
    let local = tokio::task::LocalSet::new();
    let () = local.block_on(&mut rt, async move {
        let (send, recv) = channel();
        let _ = tokio::task::spawn_local(async {
            println!("making request... ");
            let resp = reqwest::get("http://www.rust-lang.org").await?;
            println!("got resp");
            let bytes = resp.bytes().await?;
            println!("got bytes {:?}", bytes);
            send.send(()).unwrap();
            Ok::<(), Box<dyn std::error::Error>>(())
        });

        recv.await.unwrap();
    });
}
```

I was hoping this program would make the HTTP request and then terminate. Instead it gets stuck after printing "making request...".

According to @carllerche, this looks like a bug.

If I add the "full" feature to `tokio`, then the program does terminate as hoped.

I suspect `reqwest` is spawning a task and my program is failing to execute it. That actually seems kind of reasonable (if inconvenient) to me, as I never call `rt.block_on()`. However, I don't know enough about `tokio::runtime::Runtime` and `tokio::task::LocalSet` to know what to expect.

### PR DESCRIPTION
**Pull request by hawkw**
## Motivation

Currently, the only way to run a `tokio::task::LocalSet` is to call its
`block_on` method with a `&mut Runtime`, like

```rust
let mut rt = tokio::runtime::Runtime::new();
let local = tokio::task::LocalSet::new();
local.block_on(&mut rt, async {
  // whatever...
});
```

Unfortunately, this means that `LocalSet` doesn't work with the 
`#[tokio::main]`  and `#[tokio::test]` macros, since the `main` 
function is _already_ inside of a call to `block_on`.

## Solution

This branch adds a `LocalSet::run` method, which takes a future and
returns a new future that runs that future on the `LocalSet`. This
is analogous to `LocalSet::block_on`, except that it can be called in
an async context.

Additionally, this branch implements `Future` for `LocalSet`. Awaiting
a `LocalSet` will run all spawned local futures until they complete.
This allows code like

```rust
#[tokio::main] 
async fn main() {
    let local = tokio::task::LocalSet::new();

    local.spawn_local(async {
        // ...
    });

    local.spawn_local(async {
        // ...
        tokio::task::spawn_local(...);
        // ...
    });

    local.await;
}
```

I've also updated the `LocalSet` docs to show the usage with 
`#[tokio::main]` rather than with manually created runtimes, where
applicable.

Closes #1906 
Closes #1908 
Fixes #2057

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: tokio/src/task/local.rs [hunk 1]
```diff
@@ -280,12 +318,70 @@ impl LocalSet {
...
+    ///             // ...
+    ///         }).await.unwrap();
+    ///         // ...
+    ///     });
+    /// }
+    /// ```
+    ///
+    /// [`spawn_local`]: fn.spawn_local.html
+    /// [awaiting the local set]: #awaiting-a-localset
+    pub fn run<F>(&self, future: F) -> LocalFuture<F>
```
Comment (carllerche) [inline tokio/src/task/local.rs]: Maybe we should change this to `run_until` and could this be an `async fn`?

**Comment by carllerche** [comment]
Should `LocalSet::block_on` be deprecated in favor of these fns?

**Comment by carllerche** [comment]
You should also merge master and push (which will unblock CI).

Note that we have discovered that `cargo fmt` is broken. See how the rustfmt CI script has been updated.

**Comment by hawkw** [comment]
@carllerche I believe I've addressed the review feedback!

One thought I had is that it _might_ be worth having an `async fn run_until_idle(&self)` on `LocalSet` that has the behaviour of `.await`ing a LocalSet but allows it to be reused afterwards. I'm not sure if this is worth having since it creates two separate APIs for the same behavior, and being able to `.await` a local set directly is nice. Any thoughts?

**Comment by Matthias247** [comment]
FWIW, `LocalSet` in general and 

> Additionally, this branch implements Future for LocalSet. Awaiting
a LocalSet will run all spawned local futures until they complete.

in particular look more and more on what we might want to do with #1879.

I think we can try to unify these at some point. It seems like the `LocalSet` use-cases could be implemented along `runtime.block_on(task::scope(spawn_stuff_here))`, where all tasks inside the scope are restricted to run a single thread (e.g. through options which are passed to a scope?)

**Pull request merged**